From 6dd1e673ca5577fa8e4c41dfc8542301c591af36 Mon Sep 17 00:00:00 2001
From: Pete Vander Giessen <petevg@gmail.com>
Date: Wed, 8 Jun 2016 16:02:15 -0400
Subject: [PATCH] BIGTOP-2467 Zookeeper Puppet script does not setup Zookeeper
 node ids correctly

Zookeeper wants its peer ids to be consistent across nodes, but the
current puppet script auto-generates ids for nodes, based on their index
in a list. This breaks Zookeeper quorum.

We fix the issue by requiring that puppet take a list comprising
explicit ids and ip addresses, rather than just a list of ip addresses.
---
 bigtop-deploy/puppet/hieradata/bigtop/cluster.yaml              | 2 +-
 bigtop-deploy/puppet/modules/hadoop_zookeeper/templates/zoo.cfg | 4 ++--
 2 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/bigtop-deploy/puppet/hieradata/bigtop/cluster.yaml b/bigtop-deploy/puppet/hieradata/bigtop/cluster.yaml
index f592f64..45366e3 100644
--- a/bigtop-deploy/puppet/hieradata/bigtop/cluster.yaml
+++ b/bigtop-deploy/puppet/hieradata/bigtop/cluster.yaml
@@ -147,7 +147,7 @@ qfs::common::chunkserver_client_port: "22000"
 
 hadoop_zookeeper::server::myid: "0"
 hadoop_zookeeper::server::ensemble:
-  - "%{hiera('bigtop::hadoop_head_node')}:2888:3888"
+  - ["0", "%{hiera('bigtop::hadoop_head_node')}:2888:3888"]
 hadoop_zookeeper::server::kerberos_realm: "%{hiera('kerberos::site::realm')}"
 
 # those are only here because they were present as extlookup keys previously
diff --git a/bigtop-deploy/puppet/modules/hadoop_zookeeper/templates/zoo.cfg b/bigtop-deploy/puppet/modules/hadoop_zookeeper/templates/zoo.cfg
index 69d98c4..b646e60 100644
--- a/bigtop-deploy/puppet/modules/hadoop_zookeeper/templates/zoo.cfg
+++ b/bigtop-deploy/puppet/modules/hadoop_zookeeper/templates/zoo.cfg
@@ -26,8 +26,8 @@ syncLimit=5
 dataDir=<%= @datadir %>
 # the port at which the clients will connect
 clientPort=<%= @port %>
-<% @ensemble.each_with_index do |server,idx| %>
-server.<%= idx %>=<%= server %>
+<% @ensemble.each do |idx, ip| %>
+server.<%= idx %>=<%= ip %>
 <% end %>
 # purge snapshots every day
 autopurge.purgeInterval=24
-- 
2.7.4

